### PR TITLE
test(java): Add Java integration tests

### PR DESCRIPTION
**Pull request by booniepepper**
Drafting... Currently has one naive test

## Review Checklist
- [ ] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [ ] The correct base branch is being used, if not `main`
- [ ] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: config/clients/java/template/OpenFgaApiTest.java.mustache [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import org.junit.Before;
+import org.junit.Test;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.junit.jupiter.Container;
+import org.testcontainers.junit.jupiter.Testcontainers;
+import org.testcontainers.utility.DockerImageName;
+
+import static org.junit.Assert.assertEquals;
+
+public class OpenFgaApiTest {
```
Comment (booniepepper) [inline config/clients/java/template/OpenFgaApiTest.java.mustache:16]: TODO: Check conventions on other Auth0 projects using Gradle. This is an integration test. It shouldn't be necessary to run an OpenFGA server to do  local dev.

Probably this should have a different class name, and also a different Gradle task that isn't part of `check`. (So: changes here, in github workflow, and Makefile)

File: config/clients/java/template/build.gradle.mustache [hunk 2]
```diff
@@ -39,7 +44,7 @@ ext {
     {{/swagger2AnnotationLibrary}}
     jackson_version = "2.14.1"
     jakarta_annotation_version = "1.3.5"
-    junit_version = "4.13.2"
+    junit_version = "4.13.2" // TODO: JUnit 5 (Jupiter)
```
Comment (booniepepper) [inline config/clients/java/template/build.gradle.mustache:47]: TODO: Upgrade this before investing more in tests.

File: config/clients/java/template/.github/workflows/main.yml.mustache [hunk 3]
```diff
@@ -0,0 +1,27 @@
+name: Java CI with Gradle
```
Comment (booniepepper) [inline config/clients/java/template/.github/workflows/main.yml.mustache:1]: TODO: This is CI for the `fga-java-sdk` project. I should add CI for _this_ `sdk-generator` project, too.

File: Makefile [hunk 4]
```diff
@@ -207,8 +208,12 @@ setup-new-sdk:
...
+	make run-in-docker sdk_language=java image=busybox:${BUSYBOX_DOCKER_TAG} command="/bin/sh -c 'chmod +x ./gradlew'"
+	make run-in-docker sdk_language=java image=gradle:${GRADLE_DOCKER_TAG} command="/bin/sh -c './gradlew fmt'"
 
 .PHONY: test-client-java
 test-client-java: build-client-java
-	# ... any custom test code ...
+	docker container rm --force openfga-for-java-client || true
+	docker run --detach --name openfga-for-java-client -p 8080:8080 openfga/openfga run
+	make run-in-docker sdk_language=java image=gradle:${GRADLE_DOCKER_TAG} command="/bin/sh -c './gradlew build'"
+	docker container rm --force openfga-for-java-client
```
Comment (booniepepper) [inline Makefile:219]: Other clients may want to do integ tests. Abstract this to another make target?

File: config/clients/java/template/build.gradle.mustache (hunk 2 above)
Comment (booniepepper) [reply config/clients/java/template/build.gradle.mustache:48]: For now I'll just break the integration tests out to their own test suite. That can use JUnit 5 at Gradle's built-in version.

Changing this turned out to have some weird edge cases. It didn't matter what I did to the imports in the mustache files, they'd get rewritten to `import org.junit.<Class>;` instead of whatever I put down. It seems like the OpenAPI generator is doing post-processing to add/change JUnit dependencies and not strictly coming from mustache or mustache includes.

File: config/clients/java/template/.github/workflows/main.yml.mustache (hunk 3 above)
Comment (booniepepper) [reply config/clients/java/template/.github/workflows/main.yml.mustache:1]: Moved this to a checkbox in the PR description

File: Makefile (hunk 4 above)
Comment (booniepepper) [reply Makefile]: I'll leave it for later.

File: config/clients/java/template/OpenFgaApiTest.java.mustache (hunk 1 above)
Comment (booniepepper) [reply config/clients/java/template/OpenFgaApiTest.java.mustache]: done

File: config/clients/java/template/build.gradle.mustache (hunk 2 above)
Comment (rhamzeh) [reply config/clients/java/template/build.gradle.mustache:48]: Let's discuss this

File: Makefile [hunk 5]
```diff
@@ -207,8 +208,16 @@ setup-new-sdk:
...
+test-integration-client-java: build-client-java
+	docker container rm --force openfga-for-java-client || true
+	docker run --detach --name openfga-for-java-client -p 8080:8080 openfga/openfga run
+	make run-in-docker sdk_language=java image=gradle:${GRADLE_DOCKER_TAG} command="/bin/sh -c './gradlew test-integration'"
+	docker container rm --force openfga-for-java-client
 
 .PHONY: test-client-java
 test-client-java: build-client-java
-	# ... any custom test code ...
+	make run-in-docker sdk_language=java image=gradle:${GRADLE_DOCKER_TAG} command="/bin/sh -c './gradlew build'"
```
Comment (rhamzeh) [inline Makefile:223]:
```suggestion
```

You can remove this line for now, when we do add tests we can add it.

Without this line, it will just build, so this is just repeating the build step

File: config/clients/java/template/build.gradle.mustache (hunk 2 above)
Comment (booniepepper) [reply config/clients/java/template/build.gradle.mustache:48]: Turns out... the local mustache files were not being used at all. @rhamzeh figured out the issue was related to directory structure, and it's been fixed in [16b7592](https://github.com/openfga/sdk-generator/pull/158/commits/16b7592277ef25f91872346c97a5ab520dd794e8)

File: Makefile [hunk 6]
```diff
@@ -207,7 +208,16 @@ setup-new-sdk:
 .PHONY: build-client-java
 build-client-java:
 	make build-client sdk_language=java tmpdir=${TMP_DIR}
-	make run-in-docker sdk_language=java image=gradle:${GRADLE_DOCKER_TAG} command="/bin/sh -c 'gradle fmt'"
+	make run-in-docker sdk_language=java image=busybox:${BUSYBOX_DOCKER_TAG} command="/bin/sh -c 'chmod +x ./gradlew'"
+	make run-in-docker sdk_language=java image=gradle:${GRADLE_DOCKER_TAG} command="/bin/sh -c './gradlew fmt'"
```
Comment (adriantam) [inline Makefile:212]: Would it make sense to run build gradle task in the same command?  This may improve speed by not requiring spinning up the same docker twice

File: config/clients/java/template/.github/workflows/main.yml.mustache [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+on:
+  push:
+    branches: [ main, master ]
```
Comment (adriantam) [inline config/clients/java/template/.github/workflows/main.yml.mustache:5]: we are not using master, so we can remove master from the list.

File: config/clients/go/template/git_push.sh.mustache
Comment (adriantam) [inline config/clients/go/template/git_push.sh.mustache:1]: why are we touching the go sdk?

**Review by adriantam — CHANGES_REQUESTED** [review]

File: config/clients/java/template/.github/workflows/main.yml.mustache (hunk 7 above)
Comment (booniepepper) [reply config/clients/java/template/.github/workflows/main.yml.mustache:5]: Oops, I didn't notice this. Will remove

File: Makefile (hunk 6 above)
Comment (booniepepper) [reply Makefile:212]: Oh, yes good call

```suggestion
	make run-in-docker sdk_language=java image=gradle:${GRADLE_DOCKER_TAG} command="/bin/sh -c 'chmod +x ./gradlew && ./gradlew fmt build'"
```

File: config/clients/go/template/git_push.sh.mustache
Comment (booniepepper) [reply config/clients/go/template/git_push.sh.mustache:1]: Will remove this from the PR

File: Makefile (hunk 6 above)
Comment (booniepepper) [reply Makefile:212]: Along these lines, calling `./gradlew` at all forces a download, so I should avoid calling the [Gradle wrapper](https://docs.gradle.org/current/userguide/gradle_wrapper.html) altogether and use `gradle` from the Docker image.

File: config/clients/go/template/git_push.sh.mustache
Comment (booniepepper) [reply config/clients/go/template/git_push.sh.mustache:1]: Removed in [2080d74](https://github.com/openfga/sdk-generator/pull/160/commits/2080d74e72d5b19f242c1bcaee3585a9fc93fd00)

File: config/clients/java/template/.github/workflows/main.yml.mustache (hunk 7 above)
Comment (booniepepper) [reply config/clients/java/template/.github/workflows/main.yml.mustache]: Removed in [2080d74](https://github.com/openfga/sdk-generator/pull/160/commits/2080d74e72d5b19f242c1bcaee3585a9fc93fd00)

File: Makefile (hunk 6 above)
Comment (booniepepper) [reply Makefile]: Consolidated these in [2080d74](https://github.com/openfga/sdk-generator/pull/160/commits/2080d74e72d5b19f242c1bcaee3585a9fc93fd00)

File: Makefile [hunk 8]
```diff
@@ -207,7 +208,14 @@ setup-new-sdk:
 .PHONY: build-client-java
 build-client-java:
 	make build-client sdk_language=java tmpdir=${TMP_DIR}
-	make run-in-docker sdk_language=java image=gradle:${GRADLE_DOCKER_TAG} command="/bin/sh -c 'gradle fmt'"
+	make run-in-docker sdk_language=java image=gradle:${GRADLE_DOCKER_TAG} command="/bin/sh -c 'chmod +x ./gradlew && gradle fmt build'"
+
+.PHONY: test-client-java
```
Comment (adriantam) [inline Makefile:213]: Why do we have two .PHONY with test-client-java (line 213 and line 220)?

Also, will be good to update test-all-clients and build-all-clients with java so that the CICD will also exercise this as well.

**Review by adriantam — COMMENTED** [review]
We will want to update test-all-clients and build-all-clients in Make

File: Makefile (hunk 8 above)
Comment (booniepepper) [reply Makefile:213]: Oops this .PHONY is incorrect

```suggestion
.PHONY: test-integration-client-java
```

**Review by adriantam — APPROVED** [review]

File: .github/workflows/main.yaml [hunk 9]
```diff
@@ -168,3 +168,56 @@ jobs:
...
+      #     GITHUB_ORG_ID: ${{ secrets.JAVA_SDK_GITHUB_ORG_ID }}
+      #     GITHUB_REPO_ID: ${{ secrets.JAVA_SDK_GITHUB_REPO_ID }}
+      #     SSH_KEY: ${{ secrets.JAVA_SDK_SSH_KEY }}
+      #     SDK_PATH: clients/fga-java-sdk
+      #     KNOWN_HOSTS: ${{secrets.KNOWN_HOSTS}}
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+      
+      # TODO: Don't run this twice. For some reason, the initial run of
```
[the remark this answers is not in the document]
Comment (booniepepper) [reply .github/workflows/main.yaml:203]: https://github.com/openfga/sdk-generator/issues/165
